Add "Buy BSQ" button next to trade fee selector

Jakub-CZ opened the pull request:
Fixes #4579 

New button is added in the _CREATE OFFER_ and _TAKE OFFER_ tabs. When clicked, it shows a popup and then brings the user to the Sell Offer Book (_SELL BITCOIN_), with BSQ preselected in the currency filter.

The button is visible only when the fee currency toggles are visible (relevant in the _CREATE OFFER_ tab).

## Screenshots

start creating or take any offer ->

![Buy BSQ 1](https://user-images.githubusercontent.com/16313562/103590775-04d2b480-4eef-11eb-8337-5a9dfe643493.png)

-> click "BUY BSQ" ->

![Buy BSQ popup](https://user-images.githubusercontent.com/16313562/103590785-069c7800-4eef-11eb-8681-f30f69e604b5.png)

-> click "BUY BSQ" in popup ->

![image](https://user-images.githubusercontent.com/16313562/103597084-9fd38a80-4eff-11eb-9565-f1b05f8027c9.png)


## Situations when the button is clicked (unchecked are still buggy):

- [x] user is creating/taking offer to _BUY BTC_, and under _SELL BTC_ the tab "_CREATE OFFER_" is **not** active (it may be open but not selected);
- [x] user is creating/taking offer to _BUY BTC_, and under _SELL BTC_ the active tab is "_CREATE OFFER_";
- [x] user is creating offer to _SELL BTC_;
- [x] user is taking offer to _SELL BTC_.

## TODOs from discussion below:
- [x] close the current tab instead of just navigating away from it to the Offer Book;
- [x] don't display the button when the BSQ balance is already high enough.

## Known bug:

- Certain combination of the source and destination views may cause the "_Filter by currency_" combo box to receive focus, clearing it as a result. At that point, the value "BSQ" is still in effect, but any user action after that clears the filter. (More on that in the comments below.)

## Discussion:

1. If _CREATE OFFER_ to _SELL BTC_ is already open, user will be unable to create an offer, as the button _CREATE NEW OFFER TO BUY BSQ_ will be disabled. There is no indication why. A tooltip explaining why it's disabled may be helpful.
2. Should the popup include the "_don't remind me again_" option?
3. Should we hide the button when we're creating/taking an offer to trade BTC for BSQ already?
4. Having to create an altcoin account for BSQ seems like an unnecessary step. Can we have a built-in BSQ account, ideally one that will update the address to the next unused one dynamically?

Jakub-CZ committed:
Tiny refactoring

Conza88 commented:
Can we please add link to wiki on bsq etc? 
So folks can easily see trading value of purchasing etc

Jakub-CZ commented:
I've been trying to resolve the combo box focus issue for a long time, with not much success. I tried the following:
- disable the combo box temporarily,
- disable the focus listener of the combo box temporarily, or `setFocusTraversable(false)`,
- `requestFocus()` on another element, right around the time the tab switching takes place (this resolved only some cases, mostly the 2nd item in my checkbox list above),
- and much more.

These either had no effect, or caused the combo box not to reflect `sellScreenCurrencyCode` in Preferences.

Also, when the combo box receives focus, it's probably happening too early, as indicated by the fact that the drop-down list is rendered incorrectly - it's shifted to the left:

![Buy BSQ tab select - combobox focus bug](https://user-images.githubusercontent.com/16313562/103592954-b7594600-4ef4-11eb-8669-e19da2f11593.png)

### Some options that I see:
- carefuly rework AutocompleteComboBox. Instead of relying on onFocus, use onClick; I tried that already, but then the value in the combo box wasn't in line with `sellScreenCurrencyCode` in Preferences.
- instead of clearing the value when focused, just select the contents of the edit text field; thus, the previous value will not be lost, and users can still very easily filter by typing right away.

Jakub-CZ commented:
> Can we please add link to wiki on bsq etc?
> So folks can easily see trading value of purchasing etc

This one? https://bisq.wiki/Trading_fees

Is this not enough?
![image](https://user-images.githubusercontent.com/16313562/103594500-233dad80-4ef9-11eb-8bd6-2b0db7b2cb1a.png)

Conza88 commented:
Yeah that one. Just in the pop-up I think. Guides folks to wiki if they want to know benefits etc.

I think ideally should have bsq amount somewhere more prominent.
________________________________
From: Jakub Loucký <notifications@github.com>
Sent: Tuesday, January 5, 2021 11:01:47 AM
To: bisq-network/bisq <bisq@noreply.github.com>
Cc: Conza <conza_@hotmail.com>; Comment <comment@noreply.github.com>
Subject: Re: [bisq-network/bisq] Add "Buy BSQ" button next to trade fee selector (#5055)


Can we please add link to wiki on bsq etc?
So folks can easily see trading value of purchasing etc

This one? https://bisq.wiki/Trading_fees

Is this not enough?
[image]<https://user-images.githubusercontent.com/16313562/103594500-233dad80-4ef9-11eb-8bd6-2b0db7b2cb1a.png>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/bisq-network/bisq/pull/5055#issuecomment-754319203>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AMMLWYRWCCXYG2JZMDJHI5LSYJQHXANCNFSM4VT2J6CQ>.

pazza83 commented:
@Jakub-CZ looks great.

I had a read but not sure what happens when they click the 'Buy BSQ' button?

Are they taken to the Buy > BSQ/BTC market?

Jakub-CZ commented:
Yep, except it's **Sell**, as in "selling BTC for BSQ".

I've added the final screenshot in the description at the top.

pazza83 commented:
Thanks, that is great. Hopefully will get more new users to buy some BSQ :)

chimp1984 commented on desktop/src/main/java/bisq/desktop/main/offer/OfferViewUtil.java:
```diff
@@ -33,4 +47,26 @@ public static Label createPopOverLabel(String text) {
+                    navigation.navigateTo(MainView.class, SellOfferView.class, OfferBookView.class);
+                }).show());
+
+        final VBox buyBsqButtonVBox = new VBox(buyBsqButton);
```
Why a VBox is created if there is only one element in it?
The `final` in local variables should not be used (I know much of old code have it as well, but current code style suggest to not use it as unnecessary verbose).

ripcurlx replied:
Good question - I inherited this part, but didn't looked in the exact necessity of this layout. I'll simplify this positioning.

chimp1984 commented:
I had a quick look, but did not see anything obvious which could cause the problems you described. Would need to debug locally but too busy atm...

Conza88 commented:
I really think these concerns need to be addressed before sending shiteload more people to buy bsq: https://github.com/bisq-network/bisq/issues/5046

sqrrm commented:
For the case where someone is selling BTC I think the current sale tab needs to be closed and a new one opened for buying BSQ since there can only be one BTC sell tab open. That could perhaps be a bit confusing but I think allowing for multiple create offer tabs would be more trouble. Perhaps include a warning that the current BTC sell tab will be closed in the popup as well for that case.

Auto creating a BSQ account is reasonable and something I did for the not yet completed atomic BSQ trades. It's a bit more problematic for the current model as the receive address needs to be available throughout the trade process and our current model expects 1 single address per account. I think we could create a BSQ account with the first unused BSQ address if there is no other BSQ account. That would make it somewhat easier to get started with BSQ in the current state.

@Jakub-CZ I can't give any input on the UI elements not aligning and the focus, @ripcurlx knows that best I think.

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/OfferViewUtil.java:
```diff
@@ -33,4 +47,26 @@ public static Label createPopOverLabel(String text) {
+        String buyBsqText = Res.get("shared.buyCurrency", "BSQ");
+        var buyBsqButton = new AutoTooltipButton(buyBsqText);
+        buyBsqButton.getStyleClass().add("action-button");
+        buyBsqButton.setStyle("-fx-pref-height: 20; -fx-padding: 3 8 3 8; -fx-font-size: 0.769em; -fx-border-radius: 5;");
```
Please move this into the css file and use another class instead.

Jakub-CZ replied:
done

ripcurlx reviewed (commented):
I think having on the small green button `Buy BSQ` might confuse noobs what will happen if they click on this.
Maybe it should be something more like `Get BSQ` or `How to get BSQ` 👈  @m52go ?
We could use this popup afterwards not only to push the user to buy BSQ, but also show the possibility to earn BSQ by contributing.

Maybe we could also only show the Buy BSQ button if the trade has insufficient funds to pay the trading fee with BSQ.

Regarding the combobox question. So the goal is to open the offer book to sell BTC and have all BSQ offers selected? Atm it switches to the offer book setting the focus in the currency combobox.


> For the case where someone is selling BTC I think the current sale tab needs to be closed and a new one opened for buying BSQ since there can only be one BTC sell tab open. That could perhaps be a bit confusing but I think allowing for multiple create offer tabs would be more trouble. Perhaps include a warning that the current BTC sell tab will be closed in the popup as well for that case.

That is the current behavior. It will just replace the current `Take offer` tab that was used for selling BTC in the first place. Maybe we just add a `don't show again` popup if a user is trying to take different offers to reduce confusion where the old `take offer` tab vanished to.

Jakub-CZ commented:
@sqrrm
I thought about closing the tab as well, but it might be annoying to the user if they already filled in some of the fields (the button doesn't appear until at least the volume is filled in). Adding 

Also, closing the tab may get around the UI issue I'm having right now.

---
@ripcurlx 

> Regarding the combobox question. So the goal is to open the offer book to sell BTC and have all BSQ offers selected? Atm it switches to the offer book setting the focus in the currency combobox.

Yes, that is the goal. The fact that it focuses on the currency combobox is an unintended side effect of the current design. Looks like the Offer Book view gets invalidated as a result of switching from Take Offer. I don't know why the same doesn't happed when switching from BUY BTC.

> only show the Buy BSQ button if the trade has insufficient funds

Good idea. Will look into that.

wallclockbuilder reviewed (dismissed):
UtAck

Conza88 commented:
> I think having on the small green button `Buy BSQ` might confuse noobs what will happen if they click on this.
> Maybe it should be something more like `Get BSQ` or `How to get BSQ` 👈 @m52go ?
> We could use this popup afterwards not only to push the user to buy BSQ, but also show the possibility to earn BSQ by contributing.

I think it may certainly confuse any noobs, and even old crew. Also it's like a check-out, instead of one-click amazon - you're putting up barriers for people to think about before making the trade. All else being equal, they'll likely be less inclined to go through with it. Less trades, not good for Bisq. 

If the UX process was smooth for buying BSQ, I'd be less hesitant. Definitely not when it's a battle! 

Shifting to a **(i) How to get BSQ** would not be a barrier to the final trade, but an FYI for future purposes.

cd2357 commented:
@Jakub-CZ any updates or roadblocks? Been looking forward to this PR for a long time :)

Jakub-CZ commented:
@cd2357 I'll try to get back to this during the upcoming weekend.

Jakub-CZ committed:
Move button style to bisq.css

Add link to bisq.wiki

Jakub-CZ committed:
Display button only when insufficient BSQ funds

Jakub-CZ commented:
Updates since last review:
- added link to relevant wiki page;
- button styling moved to a CSS file;
- button only displayed when not enough BSQ funds to pay for the fee.

There are still buggy scenarios where the "Filter by currency" combo box gets cleared intentionally, as a result of receiving focus. Switching to BUY BTC and then back to SELL BTC activates the desired filter (BSQ).

This needs more debugging and fixing.

cd2357 commented:
Tried it out, looks promising.

You're right, it can get tricky with the closing of tabs, updating states of UI components, etc.

To keep things simple, it might be enough to not do any UI transitions and navigation to the "Sell BTC / Buy BSQ" view, but instead to just open the "Buy BSQ" popup. If the user is interested, he will follow the links and read into it. The popup's button could then link to [the step-by-step guide on how to go about and buy some BSQ](https://bisq.wiki/Paying_trading_fees_with_BSQ#Acquire_BSQ).

ripcurlx commented:
@Jakub-CZ Are you still on this PR?

Jakub-CZ commented:
> @Jakub-CZ Are you still on this PR?

Sorry, I didn't have time to get back to it. The UI bugs are difficult to iron out.

If anyone wants to take it over and finish it, I wouldn't mind.

ripcurlx commented:
> > @Jakub-CZ Are you still on this PR?
> 
> Sorry, I didn't have time to get back to it. The UI bugs are difficult to iron out.
> 
> If anyone wants to take it over and finish it, I wouldn't mind.

I'll give it a shot if I can iron out the remaining issues of this PR.

ripcurlx commented:
Just re-based on top of current master.

ripcurlx commented:
Although this is kind of repetitive, I see following open points/issues:
- [ ] 1. If `Create offer` is already open it will focus in the `Sell BTC` tab the Filter by Currency dropdown, instead of pre-selecting `BSQ`
<img width="1586" alt="Bildschirmfoto 2021-09-30 um 11 28 50" src="https://user-images.githubusercontent.com/170962/135427076-07da2982-8f78-41e0-ac5f-efb61e1a684a.png">

- [x] 2. Just showing `Buy BSQ` IMO is keeping the user in the void, what will happen when clicking on it.
<img width="1342" alt="Bildschirmfoto 2021-09-30 um 11 31 51" src="https://user-images.githubusercontent.com/170962/135427281-92f9f6b8-3338-4406-b56c-f457ad9d7f18.png">
I think it needs at least an additional help link as suggested in one of @pedromvpg 's screens
<img width="888" alt="Bildschirmfoto 2021-09-30 um 11 33 01" src="https://user-images.githubusercontent.com/170962/135427423-3b4c3cf4-76da-4dbf-89a2-5fb0950026c0.png">

- [x] 3. Having `Create Offer` open in `Sell BTC` section, will close the `Create Offer` Tab when clicking on `Buy BSQ` in the `Buy BTC > Take offer` or `Sell BTC > Create Offer` section.
**Before**
![Bildschirmfoto 2021-09-30 um 11 49 22](https://user-images.githubusercontent.com/170962/135429873-179b36f8-d6a3-4756-8878-944e5ec11db6.png)

**After clicking on `Buy BSQ`**
![Bildschirmfoto 2021-09-30 um 11 40 51](https://user-images.githubusercontent.com/170962/135428625-549ca806-37d9-4c1d-9267-414b4dcd3a57.png)

- [x]  4. Don't show `Buy BSQ` when trying to buy BSQ
![Bildschirmfoto 2021-09-30 um 11 44 34](https://user-images.githubusercontent.com/170962/135429149-07d65f7f-178c-45e5-a8bd-f45ff0aefb21.png)

- [x] 5. Process of creating a BSQ account is unclear for new users
When trying to buy BSQ user is guided to create a matching payment account.
![Bildschirmfoto 2021-09-30 um 11 54 29](https://user-images.githubusercontent.com/170962/135430817-8c7514a5-a2dc-4708-aa3a-87358a47e409.png)
Clicking on `Go to "Account"` will open the account tab in the wrong section and if the user manages to understand that she needs to go to `Altcoin Accounts` and search for BSQ. Where do I get the BSQ address from?
![Bildschirmfoto 2021-09-30 um 11 57 27](https://user-images.githubusercontent.com/170962/135431218-32fafe9e-dddd-4e2e-9002-60e39aaf4801.png)
I do think as suggested by @Jakub-CZ that we should automatically create a BSQ Altcoin account and provide the user with the information where she can find the BSQ wallet later on for confirming a trade. 
We should also add in the `Confirm Payment Receipt` screen similar to the `Start payment` screen a button link to the BSQ wallet. In this case maybe something like `Show Payment in BSQ wallet`.

@Jakub-CZ Did I miss any open points?
@Conza88 @pazza83 Any other open points from your point of view that need to be addressed in this PR?

ripcurlx commented:
@5. Something like this for confirming the payment received.
<img width="1342" alt="Bildschirmfoto 2021-09-30 um 12 22 11" src="https://user-images.githubusercontent.com/170962/135438180-ba3f1184-a5f6-42a3-91a5-1b1eac78ab2a.png">
<img width="1342" alt="Bildschirmfoto 2021-09-30 um 12 22 19" src="https://user-images.githubusercontent.com/170962/135438214-6f00a036-0046-450d-b336-44789b61c4df.png">

@m52go Any ideas for the perfect text?

I wouldn't confirm automatically right now, so users are aware of the BSQ wallet and it is less a risky addition.

ripcurlx committed:
Add Button to confirm payment in BSQ wallet

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -480,6 +480,11 @@ createOffer.triggerPrice.tooltip=As protection against drastic price movements y
+createOffer.buyBsq.popupMessage=Trading fees are paid to fund the Bisq DAO when making or taking an offer. Fees can be paid in BSQ or BTC.\n\n\
+  Bisq encourages traders to use BSQ by offering a discount of approximately 50%, since BSQ fees help fund Bisq's development. \
+  This discount varies as the BSQ/BTC rate fluctuates. To maintain the 50% discount target, trading fees are updated every cycle as necessary.\n\n\
+  For more details see [HYPERLINK:https://bisq.wiki/Trading_fees]
```
```suggestion
createOffer.buyBsq.popupMessage=Trading fees are paid to fund the Bisq DAO. Fees can be paid in BSQ or BTC.\n\n\
  BSQ fees directly help fund Bisq's development, so Bisq encourages traders to use BSQ by offering a 50% discount on trading fees. \
  This discount varies as the BSQ/BTC rate fluctuates. To maintain the 50% discount target, trading fees are updated every cycle as necessary.\n\n\
  For more about fees, see [HYPERLINK:https://bisq.wiki/Trading_fees]. For more about the Bisq DAO, see [HYPERLINK:https://bisq.wiki/Introduction_to_the_DAO#The_Bisq_DAO].
```

Just made some wording changes and added a link to the DAO introduction on the wiki.

m52go commented:
I recommend adding more context to this area in addition to the "Learn more" link. Some people still perceive BSQ to be a shitcoin, so having something short like...

"BSQ is colored BTC that helps fund Bisq developers. <a href="#">Learn More</a>" 

...next to the `BUY BSQ` button could help a lot.

![Screenshot from 2021-10-01 13-01-42](https://user-images.githubusercontent.com/735155/135659436-f9409020-5d49-46a3-a62c-3fdbc776a7aa.png)

Conza88 commented:
Fark yes to the learn more and context setting gents 👏👌🎯. 

Also very good point about coloured Bitcoin. 
Hence why I've commented elsewhere, we need BSQ to be separate compared to the shitcoins. We literally put it in that category (oof). 

But yeah, great first step 👍.

pazza83 commented:
> @Conza88 @pazza83 Any other open points from your point of view that need to be addressed in this PR?

@ripcurlx thanks for this and the useful screen shots showing the flow.

> 1. If Create offer is already open it will focus in the Sell BTC tab the Filter by Currency dropdown, instead of pre-selecting BSQ

That seems like it would be a little confusing. Any way to get the user straight to the Sell BTC / Buy BSQ screen?

> 2. Just showing Buy BSQ IMO is keeping the user in the void, what will happen when clicking on it.

I Agree, @m52go's solution https://github.com/bisq-network/bisq/pull/5055#issuecomment-932399997 looks good 

> 3. Having Create Offer open in Sell BTC section, will close the Create Offer Tab when clicking on Buy BSQ in the Buy BTC > Take offer or Sell BTC > Create Offer section.

I think this seems ok for it to close.

4. Don't show Buy BSQ when trying to buy BSQ

I think the user should be pushed towards the take offer page, not the make offer page, (more immediate)

> 5. Process of creating a BSQ account is unclear for new users. When trying to buy BSQ user is guided to create a matching payment account.

Bisq handles this well with pop up to create an account.

> Clicking on Go to "Account" will open the account tab in the wrong section and if the user manages to understand that she needs to go to Altcoin Accounts and search for BSQ. Where do I get the BSQ address from?

Maybe pop up text can be updated to include a link to the wiki and mention it is an new altcoin account. @Conza88 previously felt BSQ should have it's own tab. I think this is good but one step at a time :)

> I do think as suggested by @Jakub-CZ that we should automatically create a BSQ Altcoin account and provide the user with the information where she can find the BSQ wallet later on for confirming a trade.

Agreed

> We should also add in the Confirm Payment Receipt screen similar to the Start payment screen a button link to the BSQ wallet. In this case maybe something like Show Payment in BSQ wallet.

Agreed

ripcurlx committed:
Improve wording

Co-authored-by: m52go <735155+m52go@users.noreply.github.com>

ripcurlx commented:
> That seems like it would be a little confusing. Any way to get the user straight to the Sell BTC / Buy BSQ screen?

That is what it should do - I just pointed out the UI error with the Combox of the current implementation.

ripcurlx committed:
Not show buy bsq hint for buy bsq offers

ripcurlx committed:
Move utility methods from GUIUtil to OfferViewUtil

ripcurlx committed:
Refactor auto-closing of sell BTC offer view

ripcurlx commented:
@Jakub-CZ Was there a reason why you wanted to close the `Create Offer` view when switching the user to the Sell Bitcoin (Buy BSQ) screen?

ripcurlx committed:
Not auto-close create offer view

ripcurlx commented:
Now with more context. Clicking on `Learn More` opens the popup, clicking on `Buy BSQ` brings you to the Buy BSQ offer book.
![Bildschirmfoto 2021-10-06 um 12 13 35](https://user-images.githubusercontent.com/170962/136184208-95d73d64-9645-4ab9-8c05-4a5866082b02.png)
![Bildschirmfoto 2021-10-06 um 12 13 40](https://user-images.githubusercontent.com/170962/136184219-96b9efe1-92fc-4437-9667-cbd1570bfa43.png)
![Bildschirmfoto 2021-10-06 um 12 14 08](https://user-images.githubusercontent.com/170962/136184223-8e2451ab-85f3-4d14-b400-221dc6e28a5e.png)

ripcurlx committed:
Add more context to buy BSQ button

Jakub-CZ commented:
@ripcurlx:

I think there's no reason to close `Create Offer (to buy BTC)`.

The suggestion was to close `Create Offer (to sell BTC)`. There were two reasons:
1. If we leave it open then the user won't be able to create a new offer to buy BSQ. (also noted by @sqrrm)
2. I was hoping that closing the tab programmatically would get around the UI bug (unwanted focus on the currency combo box).

pazza83 commented:
Hi @ripcurlx 

Should this pop up also include a link to: https://bisq.wiki/Trading_BSQ

![136184219-96b9efe1-92fc-4437-9667-cbd1570bfa43](https://user-images.githubusercontent.com/66367547/136636162-fd690c58-2ced-49d6-8369-544515b6c080.png)

Conza88 commented:
> @Conza88 previously felt BSQ should have it's own tab. I think this is good but one step at a time :)

Yep ha. All in bolded below. 

_"8. Improve BSQ overall

Implement Segwit for BSQ
**[Improve BSQ account creation & purchase process](https://github.com/bisq-network/bisq/issues/5046)**
Improve BSQ accessibility through OTC market and bonded reputation
Add Low priority payment type to BSQ
BSQ balance on wallet info screen includes merit BSQ
BSQ Fees equivalent not correctly calulated into BTC in info tooltip
Very poor UX with buying BSQ. BSQ should not be in the 'alt-coin' camp either imo. Coloured BTC is better. Make more present/clear in app (without implying its better than BTC)."_

chimp1984 commented:
FYI @sqrrm 's atomic BSQ-BTC swap will be ready soon...

ripcurlx commented:
> FYI @sqrrm 's atomic BSQ-BTC swap will be ready soon...

Even better :smile: That will speed-up the BSQ purchase process even more.

ripcurlx committed:
Improve informational popup

ripcurlx commented:
Finally I found time to continue on this PR. Unfortunately it isn't possible to set a tooltip for a disabled node (mouse events are disabled), so I needed to wrap it into a StackPane putting a label on top to mimick this IMO standard behavior. TODO: Move into a re-usable component.

<img width="1367" alt="Bildschirmfoto 2021-10-11 um 11 19 53" src="https://user-images.githubusercontent.com/170962/136766015-0038e7c1-3213-4ed5-9213-244e624c7776.png">

ripcurlx committed:
Refactor account name generation

ripcurlx commented:
The current process is now as follows for no-BSQ-coiner.
Someone wants to buy BTC via Revolut
![Bildschirmfoto 2021-10-13 um 11 24 31](https://user-images.githubusercontent.com/170962/137107473-30235bdd-ceb7-41dd-adc5-d88568be0273.png)
She decides she want to save money and support the project and clicks on Buy BSQ
![Bildschirmfoto 2021-10-13 um 11 24 37](https://user-images.githubusercontent.com/170962/137107648-dd3d4a88-2603-4812-bb4f-75a787b11ef3.png)
She clicks on `Take offer to buy BSQ`
![Bildschirmfoto 2021-10-13 um 11 30 08](https://user-images.githubusercontent.com/170962/137107778-a3c89b16-6987-4f15-a141-c2b390a084d6.png)
She clicks on `Create Account` (Account is created and the offer is taken and opens the take offer screen). Probably it should be named `Create Account and Take Offer`. Also there shouldn't be a `Buy BSQ` hint when the user tries to buy BSQ.
She proceeds with purchasing BSQ
![Bildschirmfoto 2021-10-13 um 11 30 23](https://user-images.githubusercontent.com/170962/137107864-3aed24db-7e6b-4f8c-a5c7-fa4961c030fc.png)
When needing to confirm the BSQ transfer there is now the `Show Payment in BSQ Wallet` Button to open the BSQ wallet. Further optimization is not necessary as BSQ <>BTC atomic swap offers will be available soon.
![Bildschirmfoto 2021-10-13 um 11 32 06](https://user-images.githubusercontent.com/170962/137108584-366083d2-969b-4919-80a5-1917a1ded3e4.png)
![Bildschirmfoto 2021-10-13 um 11 32 36](https://user-images.githubusercontent.com/170962/137108854-d864b3f0-14ed-4865-b61e-bb8bf3fd7e03.png)
After completing the trade she can select BSQ in the offer creation process. (ATM the BSQ Button is not gone as it is not updated live)
![Bildschirmfoto 2021-10-13 um 11 32 51](https://user-images.githubusercontent.com/170962/137108916-ca1ab91f-b6c3-449e-8f4a-08ae330c051d.png)

@pazza83 @Conza88 

so IMO following points are still open to make this PR mergable:

- [x] Don't show `Buy BSQ` hint when trying to buy BSQ.
- [x] BSQ Button has to be updated live (on view activation) so it is not displayed anymore after successful BSQ purchase
- [ ] The Combobox UI glitch mentioned in https://github.com/bisq-network/bisq/pull/5055#issuecomment-931183094
- [x] Add informational popup if an BSQ instant account is created

ripcurlx commented:
I fixed the remaining issues except the Combobox UI glitch which I'll attend soon. @m52go Would be great to get also feedback on the wording of the newly introduced strings. Thanks!

ripcurlx committed:
Fix failing tests

ripcurlx commented:
- [x] Add popup after account creation to improve UX for new users

ripcurlx committed:
Add informational popup after account creation

ripcurlx commented:
@m52go Could you please have a look at the strings used in this popup? Thanks!

<img width="1411" alt="Bildschirmfoto 2021-10-15 um 11 05 03" src="https://user-images.githubusercontent.com/170962/137463103-619ef596-739e-4a02-9045-8f3e971678d2.png">

m52go commented:
I was hoping to review the strings in context on regtest but don't see the new `Buy BSQ` button when making/taking offers to buy/sell. 

Does it only show in certain special conditions?

![Screenshot from 2021-10-15 10-08-20](https://user-images.githubusercontent.com/735155/137502502-0a5e256c-af16-44a2-ae62-6eaf2f3fd4ec.png)

![Screenshot from 2021-10-15 10-08-42](https://user-images.githubusercontent.com/735155/137502509-89cfc500-2045-4a29-ab49-a91050242a16.png)

ripcurlx commented:
> I was hoping to review the strings in context on regtest but don't see the new `Buy BSQ` button when making/taking offers to buy/sell.
> 
> Does it only show in certain special conditions?
> 
> ![Screenshot from 2021-10-15 10-08-20](https://user-images.githubusercontent.com/735155/137502502-0a5e256c-af16-44a2-ae62-6eaf2f3fd4ec.png)
> 
> ![Screenshot from 2021-10-15 10-08-42](https://user-images.githubusercontent.com/735155/137502509-89cfc500-2045-4a29-ab49-a91050242a16.png)

Yes, it is only displayed if you have too little BSQ to pay the trading fees. So best way is to create a new user on Regtest.

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -439,6 +442,13 @@ offerbook.info.buyAtFixedPrice=You will buy at this fixed price.
 offerbook.info.noArbitrationInUserLanguage=In case of a dispute, please note that arbitration for this offer will be handled in {0}. Language is currently set to {1}.
 offerbook.info.roundedFiatVolume=The amount was rounded to increase the privacy of your trade.
+offerbook.info.accountCreated.headline=Congratulations
+offerbook.info.accountCreated.message=You''ve just successfully created a BSQ payment account.\n\
+  Your account can be found under Account > Altcoins Accounts > {0} and your BSQ wallet under DAO > BSQ Wallet.\n\n
```
```suggestion
offerbook.info.accountCreated.message=You''ve just successfully created a BSQ payment account.\n\n\
  See details of this payment account at Account > Altcoins Accounts > {0}. See details of your BSQ wallet at DAO > BSQ Wallet.\n\n
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -439,6 +442,13 @@ offerbook.info.buyAtFixedPrice=You will buy at this fixed price.
+offerbook.info.accountCreated.tradeInstant=As you've chosen to take a BSQ instant offer a BSQ instant account was created for you. \
+  For instant trading it is required that both trading peers are online to be able \
+  to complete the trade in less than 1 hour.\n\n
+offerbook.info.accountCreated.takeOffer=You can no proceed to take this offer after closing this popup.
```
```suggestion
offerbook.info.accountCreated.takeOffer=You can now proceed to take this offer after closing this popup.
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -439,6 +442,13 @@ offerbook.info.buyAtFixedPrice=You will buy at this fixed price.
+  Your account can be found under Account > Altcoins Accounts > {0} and your BSQ wallet under DAO > BSQ Wallet.\n\n
+offerbook.info.accountCreated.tradeInstant=As you've chosen to take a BSQ instant offer a BSQ instant account was created for you. \
+  For instant trading it is required that both trading peers are online to be able \
+  to complete the trade in less than 1 hour.\n\n
```
```suggestion
offerbook.info.accountCreated.tradeInstant=You've chosen to take a BSQ instant offer, so a BSQ instant payment account was created for you. \
  Be aware that instant trades are meant to be completed within 1 hour, \
  so you should be online and available for the next 1 hour.\n\n
```

ripcurlx committed:
Add "Buy BSQ" button next to trade fee selector

 Conflicts:
	desktop/src/main/java/bisq/desktop/main/offer/OfferView.java

ripcurlx committed:
Close current SELL offer tab

TODO: fix bug with currency combobox

 Conflicts:
	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/OfferViewUtil.java

ripcurlx committed:
Apply minor refactorings

 Conflicts:
	desktop/src/main/java/bisq/desktop/main/offer/bisq_v1/takeoffer/TakeOfferView.java

ripcurlx committed:
Add tooltip for disabled create offer button

 Conflicts:
	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookView.java

ripcurlx committed:
Simplify and clean up data models

 Conflicts:
	desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java

ripcurlx committed:
Create BSQ account and take offer if not existing

 Conflicts:
	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookView.java
	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookViewModel.java
	desktop/src/main/java/bisq/desktop/util/DisplayUtils.java

ripcurlx committed:
Fix rebase merge errors

ripcurlx committed:
Fix combobox BSQ selection error

ripcurlx committed:
Fix translations

chimp1984 reviewed (dismissed):
utACK

chimp1984 commented:
I think the "BSQ is colored coin..." can be removed then the truncating with smaller width is not happening.

<img width="1200" alt="Screenshot 2021-11-05 at 01 36 06" src="https://user-images.githubusercontent.com/54558767/140439602-a5f5747c-710b-441a-a223-3992a1fd876e.png">

Those methods are a bit weird as they mix up buyer/seller sides (isBuyBsqOffer / !isBuyOffer):
```
public boolean isBuyBsqOffer() {
        return !isBuyOffer() && getOffer().getCurrencyCode().equals("BSQ");
    }
```

ripcurlx commented:
> public boolean isBuyBsqOffer() {
>         return !isBuyOffer() && getOffer().getCurrencyCode().equals("BSQ");
>     }

Yes, this is a little bit confusing because of the fact that when you buy an altcoin you actually selling Bitcoin. I can add an additional comment to make this less confusing. But actually the naming of `isBuyBsqOffer` is accurate.

ripcurlx commented:
> I think the "BSQ is colored coin..." can be removed then the truncating with smaller width is not happening.
> 
> <img alt="Screenshot 2021-11-05 at 01 36 06" width="1200" src="https://user-images.githubusercontent.com/54558767/140439602-a5f5747c-710b-441a-a223-3992a1fd876e.png">

I'll adapt the scaling behavior so that the informational box will truncate first and within the informational box the longer informational text.

chimp1984 commented:
Yes the buy/sell direction is always referred to BTC. Only in the UI we swap it for altcoins as its not intuitive otherwise.
Maybe use a different term instead of buy/sell?

ripcurlx commented:
> > I think the "BSQ is colored coin..." can be removed then the truncating with smaller width is not happening.
> > <img alt="Screenshot 2021-11-05 at 01 36 06" width="1200" src="https://user-images.githubusercontent.com/54558767/140439602-a5f5747c-710b-441a-a223-3992a1fd876e.png">
> 
> I'll adapt the scaling behavior so that the informational box will truncate first and within the informational box the longer informational text.

I've changed the behavior so it is not truncating at 1280:
![Bildschirmfoto 2021-11-08 um 10 53 55](https://user-images.githubusercontent.com/170962/140721178-d39af583-3980-4285-aed5-b7e56f3c87d0.png)
and will look as follows with the smallest width possible:
![Bildschirmfoto 2021-11-08 um 10 54 01](https://user-images.githubusercontent.com/170962/140721166-61258580-3e58-4ad5-808e-7fedfd44974a.png)

ripcurlx committed:
Improve naming of helper method

ripcurlx committed:
Simplify containers and improve truncation behavior

chimp1984 reviewed (approved):
utACK

ripcurlx reviewed (approved):
ACK - based on https://github.com/bisq-network/bisq/pull/5055#pullrequestreview-799915337
